API security (#77)
equinor/api-strategy: docs/strategy.md

- [OPA UA explained in 1 minute (movie)](https://www.youtube.com/watch?v=-tDGzwsBokY)
- [Industrial Internet Vocabulary](https://www.iiconsortium.org/vocab/index.htm)

## <a id="api-security"></a>API Security
Equinor APIs should be crafted with high emphasis on security. Code quality is a key foundation for API security. Like other software, APIs should be built by following recognized practices for software development. Examples of such practices includes code review, test driven development (TDD), automated testing, technical debt management & refactoring, etc.

### Security tools and practices

> Development teams should include practices and tools addressing security issues in their development process. Automatic vulnerability scanning and keeping dependencies updated accordingly, is strongly recommended. Teams should stay updated on [OWASP API Security Top 10](https://github.com/OWASP/API-Security), and do regular evaluation on how these security risks are mitigated.

Reviewer: ```suggestion
Development teams should include practices and tools addressing security issues in their development process. Scanning for vulnerabilities in open source dependencies is mandatory. Teams should stay updated on [OWASP API Security Top 10](https://github.com/OWASP/API-Security), and do regular evaluation on how these security risks are mitigated.
```

Reviewer 2: I would say both: 
Automatic vulnerability scanning and keeping dependencies update accordingly, is strongly recommended. Scanning for vulnerabilities in open source dependencies is mandatory.
Your revision is missing the keep updated part.

Reviewer: I agree. I left out the keeping updated part, I kind of took that as granted and implicit. I guess we can never be to specific :)

API developers should understand and utilize the security features provided by the [API Gateway](#api-gateway), like throttling, IP restrictions and HTTP header validation.

> 

Reviewer: A few suggestions (on the borderline of a TR)(wording and location can be discussed). 

```suggestion
API's shall use encrypted transport. Transport layer security (TLS) is mandatory for REST based APIs
Input data to API's shall be sanitized.
```

### Monitoring
System monitoring, including monitoring of APIs, provide the capability to detect actual or attempted attacks on IT systems. Monitoring will also ensure that our systems and APIs are being used appropriately and in accordance with API usage policies. All Equinor APIs providing classified information should be monitored. The monitoring should include detection of unusual behaviors, like changes in IP addresses, changes in number of requests or APIs being used at unusual times of day, etc.

For more information, see [UK National Cyber Security Center Monitoring guidance](https://www.ncsc.gov.uk/collection/10-steps-to-cyber-security/the-10-steps/monitoring)

### Authentication and authorization
The company identity provider (IdP) is the basis for for authentication and authorization of Equinor APIs. Hardcoded usernames and passwords shall be avoided. OAuth2 and Open ID Connect (OIDC) are the preferred and recommended protocols for authorization and authentication of HTTP-based APIs. 

Further reading:

- [The OAuth 2.0 Authorization Framework - RFC6749](https://tools.ietf.org/html/rfc6749)
- [OpenID Connect official documentation](https://openid.net/connect/)

> 

Reviewer: A few additionals for further reading

```suggestion
[OAuth 2.0 Threat Model and Security Considerations](https://tools.ietf.org/html/rfc6819)
[ OAuth 2.0 Security Best Current Practice](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-14)
```
